### ISSUE TITLE
DeprecationWarning: Tapable.plugin is deprecated. Use new API on `.hooks` instead

### ISSUE THREAD

**Issue by i8ramin** [gatsbyjs/gatsby#13775]
## Description

This used to throw a _warning_, but now throwing an **error**

```
error (node:8947) DeprecationWarning: Tapable.plugin is deprecated. Use new API on `.hooks` instead
    at GatsbyThemeComponentShadowingResolverPlugin.apply (/Projects/project/node_modules/gatsby/dist/internal-plugins/webpack-theme-component-shadowing/index.js:34:14)
    at plugins.forEach.plugin (/Projects/project/node_modules/enhanced-resolve/lib/ResolverFactory.js:276:10)
    at Array.forEach (<anonymous>)
    at Object.exports.createResolver (/Projects/project/node_modules/enhanced-resolve/lib/ResolverFactory.js:275:10)
    at ResolverFactory._create (/Projects/project/node_modules/webpack/lib/ResolverFactory.js:60:28)
    at ResolverFactory.get (/Projects/project/node_modules/webpack/lib/ResolverFactory.js:52:28)
    at NormalModuleFactory.getResolver (/Projects/project/node_modules/webpack/lib/NormalModuleFactory.js:519:31)
    at /Projects/project/node_modules/webpack/lib/NormalModuleFactory.js:165:32
    at /Projects/project/node_modules/webpack/lib/NormalModuleFactory.js:129:4
    at hooks.beforeResolve.callAsync (/Projects/project/node_modules/webpack/lib/NormalModuleFactory.js:396:5)
    at AsyncSeriesWaterfallHook.eval [as callAsync] (eval at create (/Projects/project/node_modules/tapable/lib/HookCodeFactory.js:33:10), <anonymous>:18:1)
    at AsyncSeriesWaterfallHook.lazyCompileHook (/Projects/project/node_modules/tapable/lib/Hook.js:154:20)
    at NormalModuleFactory.create (/Projects/project/node_modules/webpack/lib/NormalModuleFactory.js:377:28)
    at semaphore.acquire (/Projects/project/node_modules/webpack/lib/Compilation.js:966:18)
    at Semaphore.acquire (/Projects/project/node_modules/webpack/lib/util/Semaphore.js:29:4)
    at Compilation._addModuleChain (/Projects/project/node_modules/webpack/lib/Compilation.js:965:18)
error (node:8947) DeprecationWarning: Resolver#doResolve: The type arguments (string) is now a hook argument (Hook). Pass a reference to the hook instead.
    at Resolver.doResolve (/Projects/project/node_modules/enhanced-resolve/lib/Resolver.js:199:11)
    at Resolver.resolver.plugin (/Projects/project/node_modules/gatsby/dist/internal-plugins/webpack-theme-component-shadowing/index.js:86:23)
    at options.fn (/Projects/project/node_modules/enhanced-resolve/lib/Resolver.js:78:9)
    at AsyncSeriesBailHook.eval [as callAsync] (eval at create (/Projects/project/node_modules/tapable/lib/HookCodeFactory.js:33:10), <anonymous>:37:1)
    at Resolver.doResolve (/Projects/project/node_modules/enhanced-resolve/lib/Resolver.js:235:16)
    at resolver.getHook.tapAsync (/Projects/project/node_modules/enhanced-resolve/lib/JoinRequestPlugin.js:21:13)
    at _next42 (eval at create (/Projects/project/node_modules/tapable/lib/HookCodeFactory.js:33:10), <anonymous>:8:1)
    at _fn42 (eval at create (/Projects/project/node_modules/tapable/lib/HookCodeFactory.js:33:10), <anonymous>:31:1)
    at resolver.getHook.tapAsync (/Projects/project/node_modules/enhanced-resolve/lib/ModuleKindPlugin.js:16:31)
    at _next41 (eval at create (/Projects/project/node_modules/tapable/lib/HookCodeFactory.js:33:10), <anonymous>:23:1)
    at _fn41 (eval at create (/Projects/project/node_modules/tapable/lib/HookCodeFactory.js:33:10), <anonymous>:46:1)
    at resolver.getHook.tapAsync (/Projects/project/node_modules/enhanced-resolve/lib/AliasPlugin.js:53:11)
    at _next40 (eval at create (/Projects/project/node_modules/tapable/lib/HookCodeFactory.js:33:10), <anonymous>:38:1)
    at _fn40 (eval at create (/Projects/project/node_modules/tapable/lib/HookCodeFactory.js:33:10), <anonymous>:61:1)
    at resolver.getHook.tapAsync (/Projects/project/node_modules/enhanced-resolve/lib/AliasPlugin.js:53:11)
    at _next39 (eval at create (/Projects/project/node_modules/tapable/lib/HookCodeFactory.js:33:10), <anonymous>:53:1)
```

### Steps to reproduce

```bash
node --trace-deprecation ./node_modules/.bin/gatsby build
```

### Environment

Run `gatsby info --clipboard` in your project directory and paste the output here.

```
  System:
    OS: macOS 10.14.3
    CPU: (8) x64 Intel(R) Core(TM) i7-7820HQ CPU @ 2.90GHz
    Shell: 5.7.1 - /usr/local/bin/zsh
  Binaries:
    Node: 11.10.1 - ~/.nvm/versions/node/v11.10.1/bin/node
    Yarn: 1.12.3 - /usr/local/bin/yarn
    npm: 6.7.0 - ~/.nvm/versions/node/v11.10.1/bin/npm
  Languages:
    Python: 2.7.15 - /usr/local/bin/python
  Browsers:
    Chrome: 73.0.3683.103
    Firefox: 54.0.1
    Safari: 12.0.3
  npmPackages:
    gatsby: ^2.3.34 => 2.3.34 
    gatsby-image: ^2.0.41 => 2.0.41 
    gatsby-plugin-advanced-sitemap: 1.0.5 => 1.0.5 
    gatsby-plugin-root-import: 2.0.5 => 2.0.5 
    gatsby-plugin-webpack-bundle-analyser-v2: 1.0.10 => 1.0.10 
```

**Comment by DSchau** [comment]
@i8ramin could you provide [a reproduction](https://gatsby.dev/reproduction)?

I've seen the warning before, as well (generally when using themes?), so taking a look at the project will help us diagnose.

**Comment by vpicone** [comment]
@DSchau I'm experiencing this warning in my theme: https://github.com/carbon-design-system/gatsby-theme-carbon

**Comment by sondr3** [comment]
I am also experiencing both of these warnings in my site as well: https://github.com/sondr3/websites/. 

1. `yarn` to install packages
2. `yarn workspace portfolio build`
3. Warnings appear.

**Comment by jrestall** [comment]
I'm also getting this warning when using themes. Seems it is related to GatsbyThemeComponentShadowingResolverPlugin using the old plugin API.

Specifically these lines:
https://github.com/gatsbyjs/gatsby/blob/master/packages/gatsby/src/internal-plugins/webpack-theme-component-shadowing/index.js#L21
https://github.com/gatsbyjs/gatsby/blob/master/packages/gatsby/src/internal-plugins/webpack-theme-component-shadowing/index.js#L85

I'm not familiar with webpack plugins enough to offer the updated code but a fix similar to https://github.com/gatsbyjs/gatsby/pull/5395/commits/6a37adc978e83ee956a64803a4355fb894ce3618#diff-346c3005d97c1ca0b5efb170af1b43f6 seems in order.

**Comment by ChristopherBiscardi** [comment]
I've reproduced this on my personal site.

![image](https://user-images.githubusercontent.com/551247/58938768-7999a780-872a-11e9-8ea9-59650cac21ee.png)

yeah it's going to be a similar fix as the one you mention @jrestall. 

It is *really* bizarre that these became errors without a major webpack version update. That's not supposed to happen. Unfortunately I can not find the PR that caused this. Since the update is pretty trivial I'll go ahead and make a PR.

**Comment by pieh** [comment]
Published `gatsby@2.8.4` with the fix.
```
npm i gatsby / yarn add gatsby
```
to upgrade :)
